// @orpc/standard-server-fastify@1.14.2/dist/index.mjs
import { once } from '@orpc/shared';
import { toAbortSignal, toStandardUrl, toStandardMethod, toStandardBody, toNodeHttpBody, toNodeHttpHeaders } from '@orpc/standard-server-node';

function toStandardLazyRequest(req, reply) {
  const signal = toAbortSignal(reply.raw);
  return {
    method: toStandardMethod(req.raw.method),
    url: toStandardUrl(req.raw),
    headers: req.headers,
    body: once(async () => {
      if (req.body !== void 0) {
        return req.body;
      }
      return toStandardBody(req.raw, { signal });
    }),
    signal
  };
}

function sendStandardResponse(reply, standardResponse, options = {}) {
  return new Promise((resolve, reject) => {
    reply.raw.once("error", reject);
    reply.raw.once("close", resolve);
    const resHeaders = { ...standardResponse.headers };
    const resBody = toNodeHttpBody(standardResponse.body, resHeaders, options);
    reply.status(standardResponse.status);
    reply.headers(toNodeHttpHeaders(resHeaders));
    reply.send(resBody);
  });
}

export { sendStandardResponse, toStandardLazyRequest };
